### PR TITLE
feat(report): Add report

### PR DESCRIPTION
**Pull request by jbrundage**

### CODE REVIEW

**Comment by jbrundage** [comment]
@GordonSmith please review

**Comment by jbrundage** [comment]
image

**Comment by GordonSmith** [comment]
@jbrundage can you squash to a single commit?

**Comment by jbrundage** [comment]
@GordonSmith done

**Comment by GordonSmith** [comment]
@dcamper you can start kicking the wheels of this - I will be code reviewing over the next week.

**Comment by dcamper** [comment]
@GordonSmith Will do, though it may be Monday before I get to it.

**Comment by dcamper** [comment]
@jbrundage Just started a functional review, using a 1Kx54 sample (1K records, 54 fields).  Enclosed screenshots show some of the results.  Comments and questions:

- There are three columns of information, but the layout of second and third columns is a bit "wide."  Can we tighten them up so the values are closer to their labels?  Or is the width a function of the data appearing somewhere within it?
- The row separator used in the first column is good, but it would be better if it (or something) extended to the full width of the display.
- There are no graphs in this result, but that may be a function of the data I'm profiling at the moment.  At which point do graphs appear?
- Cardinality/Filled percentages:  Instead of putting the percentages into another column, can you instead make them parenthetical appends to the original values?  Example: "966 (~97%)"
- Filled percentage:  Can you make the value red (and maybe bold) if the percentage is <= 33%?
- Popular patterns:  If an individual pattern is too long for the line it is on, it wraps to the next line.  Can you elide it instead?

@GordonSmith Is it possible to create a hyperlink from this report back to the result data, ideally scrolling to a row in the result?

Screen Shot 2019-03-28 at 7 20 59 AM
Screen Shot 2019-03-28 at 7 28 22 AM

**Comment by dcamper** [comment]
@jbrundage @GordonSmith Please give some thought to the documentation changes as well, both within the Profile.ecl file (which is probably minimal) and to the README (which may be much more extensive).

**Comment by GordonSmith** [comment]
We can provide a link back to the "result" - but would need an ECL Watch enhancement to "goto" a specific row.

But what we can do is display a given row from the orig data easily enough (or we could display a few rows and dim out the ones before and after to give some context...  Or we could embed the results view (paged) starting at the given row.

In version 2 I was planning on having a table view with a "dp summary" row at the top (like kaggle)

**Comment by GordonSmith** [comment]
(there is no "graph" as I wanted to get a v1 out sooner and then enhance it over time)

**Comment by dcamper** [comment]
@GordonSmith All sounds good.  Regarding the link back to the results, I was just thinking of how a user might want to explore these results, and exposing the underlying data came to mind.  As long as the data can be shown, we should be good.  Thanks!

**Comment by GordonSmith** [comment]
(You would need to keep row numbers in the profile data for the show sample option?)

**Comment by jbrundage** [comment]
@dcamper @GordonSmith I've amended this PR to include several of Dan's requests. 
image

> There are three columns of information, but the layout of second and third columns is a bit "wide." Can we tighten them up so the values are closer to their labels? Or is the width a function of the data appearing somewhere within it?
The first column is a bit wider now and the second column's labels were tightened up.

> The row separator used in the first column is good, but it would be better if it (or something) extended to the full width of the display.
The rows have alternating background colors now.

> Cardinality/Filled percentages: Instead of putting the percentages into another column, can you instead make them parenthetical appends to the original values? Example: "966 (~97%)"
Done

> Filled percentage: Can you make the value red (and maybe bold) if the percentage is <= 33%?
Waiting on this PR: https://github.com/hpcc-systems/Visualization/pull/3215

> Popular patterns: If an individual pattern is too long for the line it is on, it wraps to the next line. Can you elide it instead?
Done (but the full text will display on hover once this PR is merged: https://github.com/hpcc-systems/Visualization/pull/3215)

**Comment by dcamper** [comment]
@jbrundage Great changes!  I like them a lot.

The only thing I can see is perhaps better horizontal alignment on the values shown in the second and third columns between the rows (see enclosed screenshot).

Great work!

Screen Shot 2019-04-02 at 8 37 49 AM

**Comment by jbrundage** [comment]
@dcamper @GordonSmith I've amended this PR with improvements to the horizontal alignment

**Comment by dcamper** [comment]
@jbrundage Good improvements!  Is it possible to align the values in the third column between rows?  What is the significance of blue values?

image

**Comment by dcamper** [comment]
@GordonSmith What do you think about handling multiple profiles in the same workunit?  We may have talked about that before, but I've forgotten the answer.

Here is an example in Dataland:  http://10.173.14.207:8010/?Wuid=W20190409-143941&Widget=WUDetailsWidget#/stub/Summary

**Comment by jbrundage** [comment]
@dcamper The blue text and the 3rd column text alignment were styling bugs (they're fixed now)

**Comment by dcamper** [comment]
@jbrundage Looks great.

So the current rule is that if a cardinality breakdown is available, it is shown instead of popular patterns?

**Comment by jbrundage** [comment]
@dcamper Yes, here's the logic flow for that third column:
- if `cardinality_breakdown.Row.length > 0` then it shows the "Cardinality Breakdown" table
- else if `is_numeric === true` it shows the "Quartile Candlestick"
- else if `popular_patterns.Row.length > 0` it shows the "Popular Patterns" table
Is it possible for any of those 3 conditions to evaluate to true for one (or more) attributes?

**Comment by dcamper** [comment]
@jbrundage You could have an integer attribute (as in, the incoming data type was an integer or unsigned) with 40 unique values.  That would test true for all three of your tests.  Is that what you're asking about?

**Comment by jbrundage** [comment]
@dcamper Yes, in that case, maybe we should give each of those 3 potential widgets their own dedicated column (if the result set calls for it)

**Comment by jbrundage** [comment]
@dcamper I've amended this PR with that change. Column 3 has been broken out into columns 3, 4 and/or 5:
3) Cardinality Breakdown
4) Popular Patterns
5) Quartile Candlestick

This means that column 1 and 2 potentially have less real estate, but I've slightly restyled them to fit a bit better.

**Comment by dcamper** [comment]
@jbrundage Great change.  The callers should be able to see as much information as possible, and this helps considerably.

@GordonSmith Where are you on the code review?

**Comment by GordonSmith** [comment]
@dcamper let me know once your happy to cut an initial release and I will do the code review...

**Comment by dcamper** [comment]
@GordonSmith I think I'm good with the features.  They really look good.

File: report/res/index.css [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+.report-col-count-5 .ddCell:nth-child(10n+6),
+.report-col-count-5 .ddCell:nth-child(10n+7),
+.report-col-count-5 .ddCell:nth-child(10n+8),
+.report-col-count-5 .ddCell:nth-child(10n+9),
+.report-col-count-5 .ddCell:nth-child(10n+10) {
+    background:#F3F3F3;
+    -webkit-box-shadow: 0px 0px 0px 6px rgba(243,243,243,1);
+    -moz-box-shadow: 0px 0px 0px 6px rgba(243,243,243,1);
+    box-shadow: 0px 0px 0px 6px rgba(243,243,243,1);
+}
```
Comment (GordonSmith) [inline report/res/index.css]: What does this css do?

File: report/res/index.html [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    <link href="./index.css" rel="stylesheet">
+    <link rel="stylesheet" href="https://unpkg.com/@hpcc-js/common@2.0.0/font-awesome/css/font-awesome.min.css">
+    <script src="https://unpkg.com/@hpcc-js/loader@2.2.6/dist/index.min.js"></script>
+    <script>
+        var espUrl = window.location.href;
+        var debugWuid = window.location.search.split("?")[1];
+        if (debugWuid) {
+            espUrl = "http://play.hpccsystems.com:8010/WsWorkunits/res/" + debugWuid + "/report/res/index.html";
+        }
+        var require = window["@hpcc-js/loader"].unpkg();
```
Comment (GordonSmith) [inline report/res/index.html]: I think this loader will essentially always use the latest packages - may want to switch so its locked to a version?
Also may want to switch to using our CDN server.

File: report/runTest.ecl [hunk 3]
```diff
@@ -0,0 +1,9 @@
+IMPORT ^ as DataPatterns;
+
+filePath := 'heart_disease_uci.csv';
+
+ds := DATASET(filePath, RECORDOF(filePath, LOOKUP), CSV);
+
+profileResults := DataPatterns.Profile(ds);
+
+OUTPUT(profileResults, ALL, NAMED('profileResults'));
```
Comment (GordonSmith) [inline report/runTest.ecl]: Will probably remove from release version?

File: report/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+    secondaryFontSize: 14,
+    primaryColor: "#494949",
+    secondaryColor: "#DDD",
+    offwhiteColor: "#FBFBFB",
+    blueColor: "#1A99D5",
+    redColor: "#ED1C24",
+    fillRateRedThreshold: 33
+};
+
+export class Test extends Grid {
```
Comment (GordonSmith) [inline report/src/index.ts]: Might want to rename to "Report"

File: report/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+
+    enter(domNode, element) {
+        this._fixedHeight = this._data.length * config.rowHeight;
+        domNode.style.height = this._fixedHeight + "px";
+        this.height(this._fixedHeight);
+        super.enter(domNode, element);
+        const statsDataWidth = this.calcStatsWidgetDataColumnWidth();
+
+        this._showBreakdownColumn = this._data.filter(row=>row.cardinality_breakdown.Row.length > 0).length > 0;
+        console.log('this._showBreakdownColumn', this._showBreakdownColumn);
```
Comment (GordonSmith) [inline report/src/index.ts]: remove `console.log` calls

File: report/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+        this._showPopularPatternsColumn = this._data.filter(row=>row.popular_patterns.Row.length > 0).length > 0;
+        console.log('this._showPopularPatternsColumn', this._showPopularPatternsColumn);
```
Comment (GordonSmith) [inline report/src/index.ts]: remove console.log calls

File: report/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+        this._showQuartileColumn =  this._data.filter(row=>row.is_numeric).length > 0;
+        console.log('this._showQuartileColumn', this._showQuartileColumn);
```
Comment (GordonSmith) [inline report/src/index.ts]: remove console.log calls

File: report/src/index.ts [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+        let colCount = 3;
+        this._data.forEach((row, i) => {
+            const cc = this.enterDataRow(row, i, {statsDataWidth});
+            if(cc > colCount){
+                colCount = cc;
+            }
+        });
+        element.classed("report-col-count-"+colCount, true);
+
```
Comment (GordonSmith) [inline report/src/index.ts]: Spurious blank line

File: report/src/index.ts [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+                    background-color: ${config.offwhiteColor};
+                    float:right;
+                ">
+                    <i style="
+                        font-size:${config.secondaryFontSize}px;
+                        color:${config.blueColor};
+                    " class="fa ${row.best_attribute_type.slice(0, 6) === "string" ? "fa-font" : "fa-hashtag"}"></i>
+                    ${row.best_attribute_type}
+                </span>
+                `)
```
Comment (GordonSmith) [inline report/src/index.ts]: What is this displaying?  Should it be a widget?

File: report/src/index.ts [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+        this.adjustCellSizes();
+        return retVal;
+    }
+    adjustCellSizes() {
+        // const size = this.size();
+        // const wa = size.width * 0.618;
+        // const wb = size.width - wa;
+        // const waa = wa * 0.618;
+        // const wab = wa - waa;
+        // console.log("size", size);
```
Comment (GordonSmith) [inline report/src/index.ts]: remove commented code

File: report/res/index.css (hunk 1 above)
Comment (jbrundage) [reply report/res/index.css]: Provides the background color for the gutter between grid cells to give the appearance of a solid background color alternating between white and light grey.

File: report/src/index.ts (hunk 9 above)
Comment (jbrundage) [reply report/src/index.ts]: It's the HTMLBadge. There's a PR for it, but we were discussing whether or not it should essentially just be replaced with common_Entity.

File: report/res/index.html (hunk 2 above)
Comment (jbrundage) [reply report/res/index.html]: Is there an example you could link me to showing how the loader can be pointed to our CDN?

File: report/runTest.ecl (hunk 3 above)
Comment (jbrundage) [reply report/runTest.ecl]: Maybe it could be modified to use a small-ish inline dataset as a sort of stand-alone "hello world" example of DataPatterns?

**Comment by jbrundage** [comment]
@GordonSmith this PR has been amended with your suggested changes (except for the CDN change and the removal of `runTest.ecl`)

**Comment by dcamper** [comment]
Closing in favor of PR #31

**Pull request closed without merging**